nni/contrib/compression/pruning/slim_pruner.py

```diff
@@ -0,0 +1,197 @@
+
+        self.training_steps = training_steps
+        self.regular_scale = regular_scale
+
+        # trigger masks generation when self._current_step == self.training_steps
+        self._current_step = 0
+        # `interval_steps` and `total_times` are used by `register_trigger`.
+        # `interval_steps` is the optimize step interval for generating masks.
+        # `total_times` is the total generation times of masks.
+        self.interval_steps = training_steps
+        self.total_times: int | Literal['unlimited'] = 1
+
+        self._set_apply_method()
+
+    @classmethod
+    def from_compressor(cls, compressor: Compressor, new_config_list: List[Dict], training_steps: int,
+                        regular_scale: float = 1., evaluator: Evaluator | None = None):
+        return super().from_compressor(compressor, new_config_list, training_steps=training_steps,
+                                       regular_scale=regular_scale, evaluator=evaluator)
+
+    def _set_apply_method(self):
+        for _, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                if target_space.apply_method == 'mul':
+                    target_space.apply_method = 'slim_mul'
+                else:
+                    assert target_space.apply_method == 'slim_mul'
+
+    def _register_scaling_facotrs(self):
+        self.scaling_factors = defaultdict(dict)
+        for module_name, ts in self._target_spaces.items():
+            for target_name, target_space in ts.items():
+                if is_active_target(target_space):
+                    # TODO: add input / output
+                    if target_space.type is TargetType.PARAMETER:
+                        # TODO: here using a shrinked score to save memory, but need to test the speed.
+                        scaling_factor = torch.ones_like(target_space.target)  # type: ignore
+                        if target_space._scaler is not None:
+                            scaling_factor = target_space._scaler.shrink(scaling_factor)
+                        target_space._wrapper.register_parameter(SLIM_SCALING_FACTOR_PNAME.format(target_name),
+                                                                 torch.nn.Parameter(scaling_factor))
+                        scaling_factor = target_space._get_wrapper_attr(SLIM_SCALING_FACTOR_PNAME.format(target_name))
+                        self.scaling_factors[module_name][target_name] = scaling_factor
+                    else:
+                        raise NotImplementedError()
+
+    def _register_factors_optimization(self, evaluator: Evaluator):
+        scaling_factors = []
+        for _, target_scaling_factor in self.scaling_factors.items():
+            for _, scaling_factor in target_scaling_factor.items():
+                scaling_factors.append(scaling_factor)
+
+        if not scaling_factors:
+            return
+
+        params = [{"params": scaling_factors}]
+        optimizer = Adam(params, 1e-2)
```

Comment:
here provide patch_optimizer func for adding params_group in the evaluator,  you can use it to add new params group in the original optimizers.